spec/lang/macro/invocation_spec.lua

```diff
@@ -0,0 +1,597 @@
+
+               body_call = `($fn)($value)`
+            end
+
+            local body = block('statements')
+            table.insert(body, body_call)
+            ib[BI.IF_BLOCK.BODY] = body
+            table.insert(ifstmt[BI.IF.BLOCKS], ib)
+            ::continue::
+         end
+
+         return ifstmt
+      end
+
+      local v: integer | string | table | userdata | nil = "hello"
+
+      match!(v, {
+         function(x: integer) print(x + 10) end;
+         function(x: string) print(x:upper()) end;
+         function(x: table | userdata) print("other") end;
+         function() print("none!") end
+      })
+   ]]))
+end)
+
+describe('macro unknown type error', function()
+   it('reports unknown type from macro-built nominal without crashing', util.check_type_error([[ 
+      local macro badtype!()
+         local blk = require("teal.block")
+         local BI = blk.BLOCK_INDEXES
+
+         local function mkident(s: string)
+            local b = block('identifier')
+            b.tk = s
+            return b
+         end
+
+         -- Build: local type T = DoesNotExist
+         local lt = block('local_type')
+         lt[BI.LOCAL_TYPE.VAR] = mkident('T')
+
+         local nt = block('nominal_type')
+         nt[BI.NOMINAL_TYPE.NAME] = mkident('DoesNotExist')
+
+         local td = block('typedecl')
+         td[BI.TYPEDECL.TYPE] = nt
+
+         local newt = block('newtype')
+         newt[BI.NEWTYPE.TYPEDECL] = td
+
+         lt[BI.LOCAL_TYPE.VALUE] = newt
+
+         return lt
+      end
+
+      badtype!()
+   ]], {
+      { msg = 'unknown type DoesNotExist' },
+   }))
+end)
+
+describe('statement macro invocation (paren form)', function()
+   it('accepts bare record declaration as Statement argument', function()
+      local code = [[
+         local macro class!(b: Statement)
+            return b
+         end
+
+         class!(record MyClass
+            name: string
+         end)
+
+         local x: MyClass = { name = "ok" }
+      ]]
+      local ast, errs = tl.parse(code)
+      assert.same({}, errs)
+      local out, err = lua_gen.generate(ast, '5.4')
+      assert.is_nil(err)
+      
+      out = out:gsub("%s+", " ")
+      assert.match('local x = %s*%{ %s*name = "ok" %s*%}', out)
+   end)
+
+   it('handles commas inside Statement followed by another arg', function()
+      local code = [[
+         local macro wrap!(s: Statement, e: Expression)
+            local out = block('statements')
+            table.insert(out, s)
+            table.insert(out, `print($e)`)
+            return out
+         end
+
+         local t = {1,2,3}
+         wrap!(
+            for _, x in ipairs(t) do
+               print(x)
+            end
+         , 'ok')
+      ]]
+      local ast, errs = tl.parse(code)
+      assert.same({}, errs)
+      local out, err = lua_gen.generate(ast, '5.4')
+      assert.is_nil(err)
+      out = out:gsub('%s+', ' ')
+      assert.match('for _, x in ipairs%(%s*t%s*%) do%s*print%(x%)%s*end', out)
```

Comment:
We generally avoid complex pattern-matching in test cases like these, which make maintenance and future debugging harder. Instead, just to an exact string comparison. Also, we have util functions for testing code generation. Please use those rather than using the API directly.